lib/Doctrine/DBAL/TransactionManager.php

```diff
@@ -0,0 +1,275 @@
+
+        return $transaction;
+    }
+
+    /**
+     * @param \Doctrine\DBAL\Transaction $transaction
+     *
+     * @return void
+     *
+     * @throws \Doctrine\DBAL\ConnectionException If the transaction is stale, or belongs to another manager.
+     */
+    public function commitTransaction(Transaction $transaction)
+    {
+        if ($transaction->isActive()) {
+            // Transaction::commit() has to be called first, and in turn call this method.
+            $transaction->commit();
+
+            return;
+        }
+
+        $nestedTransaction = $this->getNestedTransaction($transaction);
+
+        if ($nestedTransaction) {
+            $nestedTransaction->commit();
+        }
+
+        $logger = $this->connection->getConfiguration()->getSQLLogger();
+
+        if (count($this->activeTransactions) === 1) {
+            $logger && $logger->startQuery('"COMMIT"');
+            $this->connection->getWrappedConnection()->commit();
+            $logger && $logger->stopQuery();
+        } elseif ($this->nestTransactionsWithSavepoints) {
+            $logger && $logger->startQuery('"RELEASE SAVEPOINT"');
+            $this->connection->releaseSavepoint($this->getNestedTransactionSavePointName());
+            $logger && $logger->stopQuery();
+        }
+
+        array_pop($this->activeTransactions);
+
+        if (! $this->connection->isAutoCommit() && ! $this->activeTransactions) {
+            $this->beginTransaction($transaction->getConfiguration());
+        }
+    }
+
+    /**
+     * @param \Doctrine\DBAL\Transaction $transaction
+     *
+     * @return void
+     *
+     * @throws \Doctrine\DBAL\ConnectionException
+     */
+    public function rollbackTransaction(Transaction $transaction)
+    {
+        if ($transaction->isActive()) {
+            // Transaction::rollback() has to be called first, and in turn call this method.
+            $transaction->rollback();
+
+            return;
+        }
+
+        $nestedTransaction = $this->getNestedTransaction($transaction);
+
+        if ($nestedTransaction) {
+            $nestedTransaction->rollback();
+        }
+
+        $logger = $this->connection->getConfiguration()->getSQLLogger();
+
+        if (count($this->activeTransactions) === 1) {
+            $logger && $logger->startQuery('"ROLLBACK"');
+            $this->connection->getWrappedConnection()->rollBack();
+            $logger && $logger->stopQuery();
+
+            if (! $this->connection->isAutoCommit()) {
+                $this->beginTransaction($transaction->getConfiguration());
+            }
+        } elseif ($this->nestTransactionsWithSavepoints) {
+            $logger && $logger->startQuery('"ROLLBACK TO SAVEPOINT"');
+            $this->connection->rollbackSavepoint($this->getNestedTransactionSavePointName());
+            $logger && $logger->stopQuery();
+        } else {
+            $this->getTopLevelTransaction()->setRollbackOnly();
+        }
+
+        array_pop($this->activeTransactions);
```

Comment:
Same here. If rolling the transaction / savepoint back fails before, the txm still holds the transaction.